fix(js-config-webpack-plugin): Fix babel internal API call failing fo…

bencoveney opened the pull request:
…r newer bable versions

## PR Checklist

Please check if your PR fulfills the following requirements:

-   [X] The commit message follows our guidelines: link tbd.
-   [ ] Tests for the changes have been added (for bug fixes / features)
-   [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

```
[X] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] tooling / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: Resolves #71 

## What is the new behavior?

## Does this PR introduce a breaking change?

```
[ ] Yes
[X] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

## Other information

Haven't added a test as this relies on changes to dependencies.

No relevant change to docs needed as far as I can tell as no functionality has changed.

I have fixed the version number of `@babel-core` as a minor version update to that is what caused the issue - The usage of `findPackageData` from `@babel/core 7.8.4` (and perhaps earlier) is incompatible with `7.2.2`.

I'm not super fluent when it comes to peer dependencies so you may want to confirm the changes there in more detail.

As mentioned in the original ticket, you may want to refer to [282f81bd676d09c1d8c48b08a7254b961eb41bed](https://github.com/babel/babel/commit/282f81bd676d09c1d8c48b08a7254b961eb41bed) in `@babel-core` as part of this PR.

I haven't included `package-lock.json` in this commit, however it is now picked up by git since its removal from `.gitignore` - let me know if you want it included.

anthonkendel commented on packages/js-config-webpack-plugin/src/JsConfigWebpackPlugin.js:
```diff
@@ -33,13 +33,25 @@ class JsConfigWebpackPlugin {
+
 		// From https://github.com/babel/babel/blob/52a569056c6008c453bf26219461655c7d0322c4/packages/babel-core/src/config/files/package.js#L15
-		const packageData = findPackageData(contextPath);
+		const packageData = getLast(findPackageData(contextPath));
```
What if checked if it is a generator function?

If it is we call it as generator function.
If it is not we call it as a normal function.

Found the code below with a quick search. There may be better ways to determine though.

https://github.com/tj/co/blob/717b043371ba057cb7a4a2a4e47120d598116ed7/index.js#L221

bencoveney replied:
Done :+1:

jantimon commented:
Thanks a lot!

Looks really good to me - @janbiasi what do you think?

jantimon commented on packages/js-config-webpack-plugin/package.json:
```diff
@@ -51,17 +51,16 @@
   },
   "devDependencies": {
-    "@babel/core": "7.2.2",
     "@babel/preset-env": "7.2.3",
```
If we use your proposed way of https://github.com/tj/co/blob/717b043371ba057cb7a4a2a4e47120d598116ed7/index.js#L221 we don't need to release a change which forces people to upgrade their babel setup.

So we can revert this change and might only upgrade babel in the dev dependencies to 7.9

bencoveney replied:
Done :+1:

jantimon commented on packages/js-config-webpack-plugin/src/JsConfigWebpackPlugin.js:
```diff
@@ -33,13 +33,25 @@ class JsConfigWebpackPlugin {
+			}
+			return result.value;
+		}
+
```
```suggestion
		/**
		 * Get the last result of an iterator
		 * @param {Generator | object} iterator
		 */
		function getLast(iterator) {
		    if (typeof iterator.next !== 'function') {
		        return iterator;
		    }
			let result = iterator.next();
			while (!result.done) {
				result = iterator.next();
			}
			return result.value;
		}

```

anthonkendel replied:
I think this looks good 👍 

Maybe clarify that it returns the parameter if it's not an iterator?

bencoveney replied:
I've gone with something like this, however I've had to invert the condition. It is sometimes valid for `iterator` to be null, in which case `typeof iterator.next !== 'function'` fails.

bencoveney replied:
I've also updated the comments/identifiers to clarify a bit what is going on 👍

bencoveney commented:
Using a variation of the suggested snippet has worked well, I can get passing tests with both versions of `@babel/core` 😁 

I've been struggling a bit against JSDoc. Having a type parameter of type `Generator<T> | T` results in a sea of red in VSCode (and the tsc lint step) because of the ambiguity.

![image](https://user-images.githubusercontent.com/1053494/77261414-478c0e80-6c86-11ea-8342-06e23ca6e543.png)

I can resolve these by adding some cast comments to the function *however* prettier is manipulating these comments so that they don't clear the errors - so I have needed to disable prettier running on the function.

The result isn't super pretty, but passes all the linting stages. Let me know if you would prefer it formatted differently - another alternative would be to `ts-ignore` the lines in question.

jantimon commented on packages/js-config-webpack-plugin/src/JsConfigWebpackPlugin.js:
```diff
@@ -33,13 +33,32 @@ class JsConfigWebpackPlugin {
+		 * @param {Generator<T> | T} valueOrIterator
+		 * @returns {T}
+		 */
+		// prettier-ignore
```
Why do we need a prettier ignore here?

bencoveney replied:
I have to add the following type assertion to keep `tsc` happy:

```
/** @type {Generator<T>} */ (valueOrIterator).next()
```

Prettier *thinks* it can improve this, and formats it without the brackets:

```
/** @type {Generator<T>} */ valueOrIterator.next()
```

However this breaks the type assertion as it now thinks:
- I am accessing the `next()` method of something that might not be a generator.
- I am casting the return value of `.next()` to be a generator instead,

Happy to implement an alternative suggestion 🙂

jantimon commented on packages/js-config-webpack-plugin/src/JsConfigWebpackPlugin.js:
```diff
@@ -33,13 +33,32 @@ class JsConfigWebpackPlugin {
+			return /** @type {T} */ (valueOrIterator);
+		}
+
 		// From https://github.com/babel/babel/blob/52a569056c6008c453bf26219461655c7d0322c4/packages/babel-core/src/config/files/package.js#L15
```
Typescript has a helper to allow you writing custom type guards:

```ts
const extractPackageData = <T extends {}>(valueOrIterator: Generator<T,T> | T): T => {
    if (!isGenerator(valueOrIterator)) {
        return valueOrIterator;
    }
    let result = valueOrIterator.next();
    while (!result.done) {
        result = valueOrIterator.next();
    }
    return result.value;
}

const isGenerator = (val: any): val is Generator => {
    // maybe we can find a better check here?
    return val && typeof val === 'object' && typeof val.next === 'function';
}
```

[Playground Link](https://www.typescriptlang.org/play?#code/MYewdgzgLgBApgDygJwIbCgBXQa1QczgBFUpUYBeGAHgBV4k4wATCGAbwF8A+ACgDdUAGwCucAPLIAklDhooIZAC4YAcSZzSiugBpa3GAB8YtAJQr6FA+wBQMezACWAMxi8AhI4jqwmhcgFhMUkZP0VTUw47BxjkOCgRZDAYQVEJaVl5RQBuaPtOPJgheJg4iBEhWCpU4Iyw5AA6XyReU1yYgHcAC0dit3cyiqgG5nA4SNsY2LhyysoUoPTQrMbmqFb2hwLphKTSmaGGmrhcgptQSFgvH3r5wKEVVDAAT3MFoSc2G5XKa0K43bJVIwABkIJgUGeAAc4CBXMCKIiYAByEAAIwAVnAMMjQeDITC4e8mogqkjkc4RGAMI5wMjTkA)

bencoveney replied:
`JsConfigWebpackPlugin.js` isn't TypeScript, but `tsc` is run as part of the linting step it looks like it will enforce whatever it finds in the JSDocs. Should `JsConfigWebpackPlugin.js` be turned into TypeScript? Or do you know if there a JSDoc alternative I could use?

bencoveney replied:
But I can definitely add your improvement to the check :+1:

anthonkendel commented:
@bencoveney Are you still working on this? If not I'm happy to help :)

bencoveney commented:
Hi @anthonkendel  - Thanks, feel free to push to the branch 👍  I've added you as a collaborator to my fork which I think would allow that? but let me know if I should add any other permissions 😄 

Hi @jantimon I've made some of the changes you've suggested but there's still a couple of questions above that I'm not sure about. I will need to do one of the following:
- Turning off some of the typesafety by using `any` in the JSDocs.
- Keeping typesafety in the the JSDocs - but with ugly cast code.
- Converting `JsConfigWebpackPlugin.js` to TypeScript.

Please let me know whatever is preferred. I've just changed the JSDocs to use `any` for now.

jantimon commented:
Sorry I guess my typescript example was to hard to understand here is the same example with jsdoc (without any): 

```
/**
		 * Return the value, unless its an interator, in which case get the last value.
		 * Works around differences in return values of @babel/core functions between versions.
		 * @template T
		 * @param {Generator<T, T> | T} valueOrIterator
		 * @returns {T}
		 */
		function getResult(valueOrIterator) {
			if (!isGenerator(valueOrIterator)) {
				return valueOrIterator;
			}
			let result = valueOrIterator.next();
			while (!result.done) {
				result = valueOrIterator.next();
			}
			return result.value;
		}

		/**
		 * Returns true if the value looks like a generator.
		 * @param {any} val
		 * @returns {val is Generator}
		 */
		function isGenerator(val) {
			return val && typeof val === 'object' && typeof val.next === 'function';
		}
```

bencoveney commented:
Thanks for clarifying - that makes sense now :+1: I didn't realise how much TypeScript syntax would be available in those JSDocs

peterekjs commented:
Probably a bit late after all the work done here 😄 As I needed to fix it immediately found out that **@babel/core** uses [gensync](https://github.com/babel/babel/blob/94447135c75a19da867f2e6a8906ce0a7e8f1ed4/packages/babel-core/src/config/files/package.js#L4) interface hence I tried to align it with that one and voila. It requires adding **gensync** package though, but the solution is pretty simple:

```js
const gensync = require('gensync')

//[...]

const packageData = gensync(findPackageData).sync(contextPath)
// needed because babels `findRelativeConfig` search just in parent directories
packageData.directories.push(packageData.filepath);

// From https://github.com/babel/babel/blob/52a569056c6008c453bf26219461655c7d0322c4/packages/babel-core/src/config/files/configuration.js#L26

// babel.config.js
const resolvedRootConfig = gensync(findRootConfig).sync(packageData.filepath, environmentName);
if (resolvedRootConfig && resolvedRootConfig.filepath) {
    return resolvedRootConfig.filepath;
}

// .babelrc.js and .babelrc
const resolvedRelativeConfig = gensync(findRelativeConfig).sync(packageData, environmentName);
if (resolvedRelativeConfig && resolvedRelativeConfig.config) {
    return resolvedRelativeConfig.config.filepath;
}
```

ernscht commented:
#71 resolved
